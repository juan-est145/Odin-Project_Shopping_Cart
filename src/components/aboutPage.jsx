import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faGithub } from "@fortawesome/free-brands-svg-icons";
import "../styles/aboutPage.css";

export function AboutPage() {
	return (
		<main className="aboutPage">
			<h1>This is a mock online store website to test the concepts of react router</h1>
			<p>The info and images from the products come from Fake Store API</p>
			<p>
				You can follow me at my github
				<a href="https://github.com/juan-est145">
					<FontAwesomeIcon icon={faGithub}></FontAwesomeIcon>
				</a>
			</p>
		</main>
	);
}
