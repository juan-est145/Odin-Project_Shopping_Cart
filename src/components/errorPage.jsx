import { Link } from "react-router-dom";
import "../styles/errorPage.css";

export function ErrorPage() {
	return (
		<main className="errorPage">
			<h1>Oops, it seems that you got lost</h1>
			<h3>
				Do not worry, you can go back to the main page by clicking this button
			</h3>
			<Link to="/">
				<button className="shopBtn">Return</button>
			</Link>
		</main>
	);
}