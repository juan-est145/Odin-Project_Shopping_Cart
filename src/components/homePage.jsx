import { Link, NavLink } from "react-router-dom";
import "../styles/homePage.css"
import motherBoardImage from "../../public/motherBoard.jpg"
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faLaptopCode, faCartShopping } from "@fortawesome/free-solid-svg-icons";

export function Header() {
	let className = "activePage";

	return (
		<header>
			<div className="headerLogo">
				<Link to="/">
					<FontAwesomeIcon icon={faLaptopCode} id="iconLogo" />
					PC Store
				</Link>
			</div>
			<nav className="headerNav">
				<NavLink
					to="/"
					className={({ isActive }) =>
						isActive ? className : null}>
					Home</NavLink>
				<NavLink
					to="shop"
					className={({ isActive }) =>
						isActive ? className : null}>
					Shop</NavLink>
				<NavLink
					to="about"
					className={({ isActive }) =>
						isActive ? className : null}>
					About</NavLink>
				<Link
					to={"order"}>
					<FontAwesomeIcon icon={faCartShopping} />
				</Link>
			</nav>
		</header>
	);
}

export function HomePage() {

	return (
		<main className="homePage">
			<img src={motherBoardImage} alt="An image of a mother board" className="mainPhoto" />
			<div className="homeText">
				<h1>Looking for hardware? You are at the right place.</h1>
				<h3>PC Store provides great hardware for a competitive price. Don&apos;t believe us? Check it out!</h3>
			</div>
			<Link to="shop">
				<button className="shopBtn">Shop now</button>
			</Link>
		</main>
	)
}
