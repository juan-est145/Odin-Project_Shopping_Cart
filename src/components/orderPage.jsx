import { useOutletContext } from "react-router-dom";
import PropTypes from "prop-types";
import "../styles/orderPage.css";

export function OrderPage() {
	const { order, setOrder } = useOutletContext();

	if (Object.entries(order).length <= 0)
	{
		return (
			<h1>Your shopping cart is empty at the moment</h1>
		);
	}

	return (
		<main>
			<section>
				{Object.entries(order).map(([id, element]) => (
					<ShopOrder key={id} shopItem={element} setOrder={setOrder}></ShopOrder>
				))}
			</section>
			<section>
				<TotalPayment order={order}></TotalPayment>
			</section>
		</main>
	);
}

function ShopOrder({ shopItem, setOrder }) {
	return (
		<div className="orderCard">
			<img src={shopItem.image} alt={`Photo of a ${shopItem.title}`} />
			<div>
				<h3>{shopItem.title}</h3>
				<p>{`${shopItem.price} x ${shopItem.quantity} = ${shopItem.price * shopItem.quantity} €`}</p>
			</div>
		</div>

	);
}

function TotalPayment({order})
{
	return (
		<h1>{`${Object.values(order).reduce((total, amount) => (
				(total.price * total.quantity) + (amount.price * amount.quantity)
		))} €`}</h1>
	)
}

ShopOrder.propTypes = {
	shopItem: PropTypes.object,
	setOrder: PropTypes.func,
}

TotalPayment.propTypes = {
	order: PropTypes.object,
}