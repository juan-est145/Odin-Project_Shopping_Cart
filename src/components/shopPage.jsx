import { Link, useOutletContext } from "react-router-dom";
import { useEffect } from "react";
import PropTypes from "prop-types";

export default function ShopPage() {
	const {items, setItems} = useOutletContext();

	useEffect(() => {
		if (items.length !== 0)
			return;
		getShopItems().then((shopItems) => {
			if (shopItems)
				setItems(shopItems);
		});
	}, [items, setItems]);

	return (
		<>
			<h1>Shop</h1>
			<h3>See something that you fancy?</h3>
			<Items items={items}></Items>
			<Link to="/">Go back</Link>
		</>
	);
}

function Items({ items }) {
	return (
		<>
			{items.map((element) => {
				return (
					<ul key={element.id}>
						<li>{element.title}</li>
						<li>{element.price}</li>
						<li>{element.category}</li>
						<li>{element.description}</li>
						<img src={element.image} alt="Something" />
					</ul>
				);
			})}
		</>
	);
}

async function getShopItems() {
	try {
		const response = await fetch('https://fakestoreapi.com/products/category/electronics', { mode: "cors" });
		const results = await response.json();
		return (results);
	} catch (error) {
		console.error("Failed to fetch items data", error);
	}
}

Items.propTypes = {
	items: PropTypes.array,
};