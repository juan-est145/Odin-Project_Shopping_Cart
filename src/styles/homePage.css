@font-face {
	font-family: "Roboto";
	src: url("../../public/roboto/Roboto-Regular.ttf");
}

@font-face {
	font-family: "Redig";
	src: url("../../public/redigFont/Redig-Regular.ttf");
}

body {
	margin: 0px;
	font-family: "Roboto";
}

#root {
	display: flex;
	flex-direction: column;
	min-height: 100vh;
}

header {
	display: flex;
	justify-content: space-between;
	padding: 1.5em;
	background-color: black;
	color: white;
	font-size: 1.4rem;
	text-wrap: nowrap;
	gap: 10px;
}

.headerLogo {
	display: flex;
	align-items: center;
	flex-grow: 5;
	color: white;
	gap: 20px;
}

.headerLogo>a {
	color: white;
	display: flex;
	font-family: "Redig";
	align-self: center;
	font-size: 2.6rem;
	gap: 10px;
}

#iconLogo {
	align-self: flex-end;
	margin-top: 12px;
}

.headerNav {
	flex-grow: 1;
	display: flex;
	justify-content: space-evenly;
	align-items: center;
	gap: 10px;
	font-size: 1.6rem;
}

.headerNav>* {
	color: white;
	text-decoration: none;
	cursor: pointer;
}

.headerNav>*:hover,
.activePage {
	text-decoration: underline;
}

main {
	display: flex;
	flex-direction: column;
	flex-grow: 1;
	text-align: center;
	justify-content: center;
	align-items: center;
	color: white;
}

.mainPhoto {
	position: absolute;
	width: 100%;
	min-height: 100%;
	object-fit: cover;
	z-index: -1;
}

.homeText {
	max-width: 50ch;
	font-size: 1.4rem;
}

.shopBtn {
	--btn-color: rgb(15, 43, 226);
	--btn-hover: rgb(10, 27, 143);
	font-size: 1.3rem;
	font-weight: bold;
	padding: 0.6em;
	border-radius: 1.4em;
	border: none;
	cursor: pointer;
	background-color: var(--btn-color);
	color: white;
}

.shopBtn:hover {
	background-color: var(--btn-hover);
}