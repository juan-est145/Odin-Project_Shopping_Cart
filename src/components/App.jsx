import { useState } from 'react';
import { Outlet } from "react-router-dom";
import { Header } from "./homePage";

function App() {
	const [items, setItems] = useState([]);
	const [order, setOrder] = useState({});
	
	return (
		<>
			<Header order={order}></Header>
			<Outlet context={{items, setItems, order, setOrder}}></Outlet>
		</>
	);
}

export default App
