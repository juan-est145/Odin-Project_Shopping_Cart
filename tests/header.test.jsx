import { describe, it, expect } from "vitest";
import { render, screen } from "@testing-library/react";
import { Header } from "../src/components/homePage";
import { BrowserRouter } from "react-router-dom";
//import App from "../src/components/App";
import { HomePage } from "../src/components/homePage";

describe("Header", () => {
	it("Renders the basic stuff", () => {
		const { container } = render(
			<BrowserRouter>
				<Header />
			</BrowserRouter>
		);
		expect(container).toMatchSnapshot();
	})
});

describe("Home page", () => {
	it("Renders the title page", () => {
		const description = {
			exact: false,
			name : "Looking for hardware? You are at the right place.",
		}
		render(
            <BrowserRouter>
                <HomePage />
            </BrowserRouter>
        );
		expect(screen.getByRole("heading", description).textContent).toMatch("Looking for hardware? You are at the right place.");
	});
});