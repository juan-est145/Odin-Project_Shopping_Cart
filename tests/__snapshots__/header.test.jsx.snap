// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Header > Renders the basic stuff 1`] = `
<div>
  <header>
    <div
      class="headerLogo"
    >
      <a
        href="/"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-laptop-code "
          data-icon="laptop-code"
          data-prefix="fas"
          focusable="false"
          id="iconLogo"
          role="img"
          viewBox="0 0 640 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M64 96c0-35.3 28.7-64 64-64l384 0c35.3 0 64 28.7 64 64l0 256-64 0 0-256L128 96l0 256-64 0L64 96zM0 403.2C0 392.6 8.6 384 19.2 384l601.6 0c10.6 0 19.2 8.6 19.2 19.2c0 42.4-34.4 76.8-76.8 76.8L76.8 480C34.4 480 0 445.6 0 403.2zM281 209l-31 31 31 31c9.4 9.4 9.4 24.6 0 33.9s-24.6 9.4-33.9 0l-48-48c-9.4-9.4-9.4-24.6 0-33.9l48-48c9.4-9.4 24.6-9.4 33.9 0s9.4 24.6 0 33.9zM393 175l48 48c9.4 9.4 9.4 24.6 0 33.9l-48 48c-9.4 9.4-24.6 9.4-33.9 0s-9.4-24.6 0-33.9l31-31-31-31c-9.4-9.4-9.4-24.6 0-33.9s24.6-9.4 33.9 0z"
            fill="currentColor"
          />
        </svg>
        PC Store
      </a>
    </div>
    <nav
      class="headerNav"
    >
      <a
        aria-current="page"
        class="activePage"
        href="/"
      >
        Home
      </a>
      <a
        href="/shop"
      >
        Shop
      </a>
      <a
        href="/about"
      >
        About
      </a>
      <a
        href="/order"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-cart-shopping "
          data-icon="cart-shopping"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 576 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 24C0 10.7 10.7 0 24 0L69.5 0c22 0 41.5 12.8 50.6 32l411 0c26.3 0 45.5 25 38.6 50.4l-41 152.3c-8.5 31.4-37 53.3-69.5 53.3l-288.5 0 5.4 28.5c2.2 11.3 12.1 19.5 23.6 19.5L488 336c13.3 0 24 10.7 24 24s-10.7 24-24 24l-288.3 0c-34.6 0-64.3-24.6-70.7-58.5L77.4 54.5c-.7-3.8-4-6.5-7.9-6.5L24 48C10.7 48 0 37.3 0 24zM128 464a48 48 0 1 1 96 0 48 48 0 1 1 -96 0zm336-48a48 48 0 1 1 0 96 48 48 0 1 1 0-96z"
            fill="currentColor"
          />
        </svg>
      </a>
    </nav>
  </header>
</div>
`;
